cmd/gossamer/commands/root.go

```diff
@@ -0,0 +1,628 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package commands
+
+import (
+	"fmt"
+
+	"github.com/ChainSafe/gossamer/lib/keystore"
+
+	cfg "github.com/ChainSafe/gossamer/config"
+
+	"github.com/ChainSafe/gossamer/internal/log"
+
+	westenddev "github.com/ChainSafe/gossamer/chain/westend-dev"
+	"github.com/ChainSafe/gossamer/dot"
+	"github.com/spf13/cobra"
+)
+
+// Package level variables
+var (
+	config = westenddev.DefaultConfig()
+	logger = log.NewFromGlobal(log.AddContext("pkg", "cmd"))
+)
+
+// Flag values for the root command which needs type conversion
+var (
+	// Base Config
+	logLevelGlobal string
+	pruning        string
+	telemetryURLs  string
+
+	// Log Config
+	logLevelCore    string
+	logLevelDigest  string
+	logLevelSync    string
+	logLevelNetwork string
+	logLevelRPC     string
+	logLevelState   string
+	logLevelRuntime string
+	logLevelBABE    string
+	logLevelGRANDPA string
+
+	// Core Config
+	// role of the node. one of: full, light or authority
+	role string
+	// validator when set, the node will be an authority
+	validator bool
+
+	// Account Config
+	// key to use for the node
+	key string
+)
+
+// Flag values for persistent flags
+var (
+	// Default accounts
+	alice   bool
+	bob     bool
+	charlie bool
+
+	// Initialization flags for node
+	chain    string
+	basePath string
+)
+
+// Default values
+const (
+	// DefaultHomeEnv is the default environment variable for the base path
+	DefaultHomeEnv = "GSSMRHOME"
+)
+
+// NewRootCommand creates the root command
+func NewRootCommand() (*cobra.Command, error) {
+	cmd := &cobra.Command{
+		Use:   "gossamer",
+		Short: "Official gossamer command-line interface",
+		Long: `Gossamer is a Golang implementation of the Polkadot Host.
+Usage:
+	gossamer --chain westend-local --alice --babe-lead
+	gossamer --chain westend-dev --key alice --port 7002
+	gossamer --chain westend --key bob --port 7003
+	gossamer --chain kusama --key charlie --port 7004
+	gossamer --chain polkadot --key dave --port 7005`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return execRoot(cmd)
+		},
+		PersistentPreRunE: func(cmd *cobra.Command, args []string) (err error) {
+			if !(cmd.Name() == "gossamer" || cmd.Name() == "init") {
+				return nil
+			}
+
+			if err := parseChainSpec(chain); err != nil {
+				return fmt.Errorf("failed to parse chain-spec: %s", err)
+			}
+
+			if err := parseBasePath(); err != nil {
+				return fmt.Errorf("failed to parse base path: %s", err)
+			}
+
+			parseAccount()
+
+			// Copy chain-spec to base-path
+			if err := copyChainSpec(config.ChainSpec, cfg.GetChainSpec(config.BasePath)); err != nil {
+				return fmt.Errorf("failed to copy chain-spec: %s", err)
+			}
+
+			if cmd.Name() == "gossamer" {
```

Comment:
does this expect command to start exactly with "gossamer"? what happens if someone runs command using something like "./bin/gossamer"?

Reply:
`cmd.Name()` will always be equal to [this](https://github.com/ChainSafe/gossamer/blob/refactor/k/cli/cmd/gossamer/commands/root.go#L76)